experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -127,63 +128,103 @@ static void markPathToCommits(llvm::DenseSet<Operation *> &markedPath,
+// Value: The value whose spec tag is used
+// handshake::ConditionalBranchOp: The branch to replicate
+// unsigned: The direction of the branch to follow
+typedef std::list<std::tuple<Value, handshake::ConditionalBranchOp, unsigned>>
+    CommitBranchList;
+// This function traverses the IR along a marked path and creates
 // a control path by replicating the branches it finds in the way. It stops
 // at commits and connects them to the newly created path with value
 // ctrlSignal
 void HandshakeSpeculationPass::routeCommitControl(
-    llvm::DenseSet<Operation *> &markedPath, Value ctrlSignal,
-    OpOperand &currOpOperand) {
-  Operation *currOp = currOpOperand.getOwner();
-  // End traversal if currOp is not in the marked path to commits
-  if (!markedPath.contains(currOp))
-    return;
-
-  // Remove operation from the set to avoid visiting it twice
-  markedPath.erase(currOp);
-
-  if (auto commitOp = dyn_cast<handshake::SpecCommitOp>(currOp)) {
-    // Connect commit to the correct control signal and end traversal
-    commitOp.setOperand(1, ctrlSignal);
-  } else if (auto branchOp = dyn_cast<handshake::ConditionalBranchOp>(currOp)) {
-    // Replicate a branch in the control path and use new control signal.
-    // To do so, a structure of two connected branches is created.
-    // A speculating branch first discards the condition in case that
-    // the data is not speculative. In case it is speculative, a new branch
-    // is created that replicates the current branch.
-
-    MLIRContext *ctx = &getContext();
-    OpBuilder builder(ctx);
-    builder.setInsertionPointAfterValue(ctrlSignal);
-
-    // The speculating branch will discard the branch's condition token if the
-    // branch output is non-speculative. Speculative tag of the token is
-    // currently implicit, so the branch input itself is used at the IR level.
-    auto branchDiscardNonSpec = builder.create<handshake::SpeculatingBranchOp>(
-        branchOp.getLoc(), currOpOperand.get() /* specTag */,
-        branchOp.getConditionOperand());
-    inheritBB(specOp, branchDiscardNonSpec);
-
-    // The replicated branch directs the control token based on the path the
-    // speculative token took
-    auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
-        branchDiscardNonSpec->getLoc(),
-        branchDiscardNonSpec.getTrueResult() /* condition */,
-        ctrlSignal /* data */);
-    inheritBB(specOp, branchReplicated);
-
-    // Follow the two branch results with a different control signal
-    for (unsigned i = 0; i <= 1; ++i) {
-      for (OpOperand &dstOpOperand : branchOp->getResult(i).getUses()) {
-        Value ctrl = branchReplicated->getResult(i);
-        routeCommitControl(markedPath, ctrl, dstOpOperand);
+    llvm::DenseSet<Operation *> &markedPath) {
+
+  // Queue for BFS
+  std::queue<std::tuple<const OpOperand &, CommitBranchList>> queue;
+
+  // Initialize the queue with the dataOut Operand of specOp (usually only one)
```

Comment:
explicitly mention fork

Reply:
We changed the implementation so now this is outdated. I resolve this conversation